### ISSUE TITLE
[BUG] Windows QZ crashing upon startup

### ISSUE THREAD

**Issue by failedethics** [cagnulein/qdomyos-zwift#1670]
**Describe the bug**
Windows QZ(mscv2019-binary-no-python) freezes upon startup with connection to Horizon 7.0AT-04. (Has also frozen with no connection while going into settings). 

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 11 Home (Installed on ROG Zephyrus M16 GU603ZM)
 - Version 22H2

**Smartphone (please complete the following information):**

**Append a debug log**

 Follow this guide https://github.com/cagnulein/qdomyos-zwift/wiki/How-do-i-get-the-debug-log-in-case-something-doesn't-work%3F
  
  
**Additional context**
[debug-Mon_Sep_25_12_29_31_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12718007/debug-Mon_Sep_25_12_29_31_2023.log)
![unnamed](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/22a48c11-94c4-40de-8274-b0ef9c56862e)

**Comment by cagnulein** [comment]
does it do the same without the zwift setting enabled in qz?

Il giorno lun 25 set 2023 alle 19:40 failedethics ***@***.***>
ha scritto:

**Comment by failedethics** [comment]
Under "Zwift Options" disabled "Zwift Treadmill Auto Inclination" and freeze still occured. 
[debug-Mon_Sep_25_13_11_09_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12718262/debug-Mon_Sep_25_13_11_09_2023.log)

**Comment by cagnulein** [comment]
I don't see any connection here, are you sure it's the right log? could you
please retry in order to be sure it's the right one?
thanks

Il giorno lun 25 set 2023 alle 20:14 failedethics ***@***.***>
ha scritto:

**Comment by failedethics** [comment]
Sure thing. It finds it, doesn't bring in any stats(set to 1MPH), then crashes and requires a force close. 
[debug-Mon_Sep_25_13_26_12_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12718334/debug-Mon_Sep_25_13_26_12_2023.log)
![Screenshot 2023-09-25 132737](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/887eee92-4ae2-405e-9f22-63ea26564cce)

**Comment by cagnulein** [comment]
got the issue, let me try to patch if from the phone (9pm here)

Il giorno lun 25 set 2023 alle 20:29 failedethics ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
i did a patch but i'm not sure 
it's building now https://github.com/cagnulein/qdomyos-zwift/actions/runs/6303842224

you will see the new msvc build at the bottom of the page when it's ready (usually 40 minutes)

let me know

**Comment by failedethics** [comment]
I used the "windows-msvc2019-binary-no-python" Same result, debug attached.
[debug-Mon_Sep_25_15_23_18_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12719054/debug-Mon_Sep_25_15_23_18_2023.log)

**Comment by cagnulein** [comment]
thanks i will check it asap, all my family have the flu and it's a
nightmare with a baby

i will update you asap
thanks

Il giorno lun 25 set 2023 alle 22:26 failedethics ***@***.***>
ha scritto:

**Comment by failedethics** [comment]
Take your time and take care of the family! It's 11pm here so not messing
with it tonight.

**Comment by cagnulein** [comment]
ok I tried to relax things, let's try this
https://github.com/cagnulein/qdomyos-zwift/actions/runs/6310955973
Let me know

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 26 set 2023 alle ore 05:56 failedethics <
***@***.***> ha scritto:

**Comment by failedethics** [comment]
Same thing, connects to the horizon then crashes. 
[debug-Tue_Sep_26_08_03_35_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12727878/debug-Tue_Sep_26_08_03_35_2023.log)

**Comment by cagnulein** [comment]
I don't understand why for another user this solution it works and for you
it doesn't.
I will try to do another approach.
I will let you know when I will be ready
Thanks for your time

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 26 set 2023 alle ore 15:06 failedethics <
***@***.***> ha scritto:

**Comment by failedethics** [comment]
I agree, it's baffling. I've sent an email to your gmail to continue the
conversation and attempt troubleshooting.

**Comment by cagnulein** [comment]
@failedethics building a new one https://github.com/cagnulein/qdomyos-zwift/actions/runs/6314216910
Let me know

**Comment by failedethics** [comment]
Sorry 
[debug-Tue_Sep_26_10_29_49_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12729425/debug-Tue_Sep_26_10_29_49_2023.log)

**Comment by cagnulein** [comment]
which windows version do you have?

**Comment by failedethics** [comment]
Device info: 

Device name	ROG
Processor	12th Gen Intel(R) Core(TM) i7-12700H   2.30 GHz
Installed RAM	16.0 GB (15.7 GB usable)
Device ID	B0ED9BA0-F160-4B69-82FB-D4680ADD5473
Product ID	00342-20881-21596-AAOEM
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display

**Comment by cagnulein** [comment]
mmm i guess the other users was with windows 10. don't you have a windows 10 machine just to test if the issue will disappear? if so we can think about a solution

**Comment by failedethics** [comment]
My gaming PC is windows 10, I would just have to move it out to the gym. I had an old lenovo yoga I was going to try but it's fried. I'll try to get the desktop out there today and will update with findings. Can you tell me/provide a direct link to the version you want me to put on the desktop?

**Comment by cagnulein** [comment]
ok great! the last one you tried https://github.com/cagnulein/qdomyos-zwift/suites/16579760991/artifacts/947253348

**Comment by failedethics** [comment]
Desktop info:

Edition Windows 10 Home
Version 22H2
Installed on ‎2/‎27/‎2021
OS build 19045.3448
Experience Windows Feature Experience Pack 1000.19044.1000.0

Device name DESKTOP-O749ESC
Processor AMD Ryzen 5 3600 6-Core Processor                 3.59 GHz
Installed RAM 15.9 GB
Device ID 4010041D-6C7C-4BCB-9FDC-1B1A0C38F259
Product ID 00325-81683-30907-AAOEM
System type 64-bit operating system, x64-based processor

**Comment by cagnulein** [comment]
yes should be very similar to the one of the other users
In case it doesn't work either, I have a last bullet to fire.

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 26 set 2023 alle ore 17:42 failedethics <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
for my reference, the last bullet will be this one https://codereview.qt-project.org/c/qt/qtconnectivity/+/392128/7/src/bluetooth/qlowenergycontroller_winrt.cpp#738

**Comment by failedethics** [comment]
Desktop worked, although incline didn't adjust but that's probably a settings issue.

**Comment by cagnulein** [comment]
Glad to hear it!
Did you enable the zwift setting in qz? because now we should check if the
python script it's working ok with this branch (this branch is for
treadmill like your that can't connect in the windows settings)


Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 26 set 2023 alle ore 18:06 failedethics <
***@***.***> ha scritto:

**Comment by failedethics** [comment]
I have "Zwift Treadmill Auto Inclination" enabled, there is also Treadmill Climb Portal and Treadmill Auto Workout. Which ones should I enable and I'll retest.

**Comment by cagnulein** [comment]
yes "Zwift Treadmill Auto Inclination" it's correct. so let's try to append
a debug log, I will check it if the python script is working

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 26 set 2023 alle ore 18:11 failedethics <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
@failedethics also check this https://github.com/cagnulein/qdomyos-zwift/issues/1675#issuecomment-1735873865

**Comment by failedethics** [comment]
Also, to confirm, in Zwift I have the option for Wahoo Kickr and wahoo tread, I selected Tread, but let me know if I should change to kickr. I let it run for 8 mins and go through 2 climbs, no reponse from incline. 
[debug-Tue_Sep_26_11_14_09_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12729875/debug-Tue_Sep_26_11_14_09_2023.log)

**Comment by cagnulein** [comment]
got the issue

`Tue Sep 26 11:14:15 2023 1695744855813 Debug: .\bluetooth.cpp void __cdecl bluetooth::debug(const class QString &) "python << ERR Traceback (most recent call last):\r\n  File \"C:\\Users\\18304\\Downloads\\windows-msvc2019-binary\\release\\output\\zwift-incline.py\", line 92, in <module>\r\n    for word in line:\r\nTypeError: 'NoneType' object is not iterable\r\n"`

@victorypoint did you ever see this?

@failedethics did you also check my previous link?

**Comment by failedethics** [comment]
Trying borderless window now

**Comment by failedethics** [comment]
[debug-Tue_Sep_26_11_38_24_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12730113/debug-Tue_Sep_26_11_38_24_2023.log)
Took it up a 7% incline in game, no response.

**Comment by cagnulein** [comment]
same error, are you able to run the python script zwift-incline.py in the
qz folder manually?

Il giorno mar 26 set 2023 alle 18:57 failedethics ***@***.***>
ha scritto:

**Comment by failedethics** [comment]
No idea how to do that

**Comment by cagnulein** [comment]
open a prompt on windows
go into the qdomyos folder

than do

.\python\\x64\\python.exe zwift-incline.py

let me know

**Comment by failedethics** [comment]
![attempt](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/c75fccec-9cf0-440a-a103-87f401bc5214)
let me know if that was done correctly

**Comment by cagnulein** [comment]
ok perfect so same issue, i will check this tomorrow
i hope we are near

Il giorno mar 26 set 2023 alle 21:03 failedethics ***@***.***>
ha scritto:

**Comment by victorypoint** [comment]
> got the issue
> 
> `Tue Sep 26 11:14:15 2023 1695744855813 Debug: .\bluetooth.cpp void __cdecl bluetooth::debug(const class QString &) "python << ERR Traceback (most recent call last):\r\n File \"C:\\Users\\18304\\Downloads\\windows-msvc2019-binary\\release\\output\\zwift-incline.py\", line 92, in <module>\r\n for word in line:\r\nTypeError: 'NoneType' object is not iterable\r\n"`
> 
> @victorypoint did you ever see this?
> 
> @failedethics did you also check my previous link?

Hi guys. Just reading this thread now. I've not encountered this crash issue @cagnulein. Let me know if you guys need me to test anything.

**Comment by cagnulein** [comment]
@victorypoint it seems to me that error is about the fact the result variable is None (so ocr can't recognize anything). Maybe we can handle it adding an if before the for line? Anyway I don't think it's the main issue of @failedethics 

@failedethics are you running in 4k resolution? also can you try to follow this https://github.com/cagnulein/qdomyos-zwift/issues/1581#issuecomment-1660774335 ?

**Comment by victorypoint** [comment]
> @victorypoint it seems to me that error is about the fact the result variable is None (so ocr can't recognize anything). Maybe we can handle it adding an if before the for line? Anyway I don't think it's the main issue of @failedethics
> 
> @failedethics are you running in 4k resolution? also can you try to follow this [#1581 (comment)](https://github.com/cagnulein/qdomyos-zwift/issues/1581#issuecomment-1660774335) ?

@cagnulein, I'll download the latest QZ build you indicated above and try to reproduce error. Standby..

**Comment by cagnulein** [comment]
thanks @victorypoint 
consider that this build is highly experimental since i'm using msvc instead of mingw but it shouldn't affect the python script . so my guess is that there is something in the @failedethics setup that we should understand

**Comment by victorypoint** [comment]
> thanks @victorypoint consider that this build is highly experimental since i'm using msvc instead of mingw but it shouldn't affect the python script . so my guess is that there is something in the @failedethics setup that we should understand

Ok. Why not have failedethics use the standard 'windows-binary' in this case?

**Comment by victorypoint** [comment]
@cagnulein, ah of course. Okay, I tested the msvc QZ build and python/PaddleOCR appears to work fine. The getincline.bat I included in this post - https://github.com/cagnulein/qdomyos-zwift/issues/1581#issuecomment-1660774335 - works fine and returns 2% so have @failedethics test that and make sure his python environment variables and set.

**Comment by failedethics** [comment]
Sorry, just catching up. 
- Resolution on Win10 device: 1920x1080
- To ensure I'm executing your request properly, I will: 

1. Right-click on the Start Button
2. Select “System” from the context menu.
3. Click “Advanced system settings”
4. Go to the “Advanced” tab
5. Click “Environment Variables…”
6. Click variable called “Path” and click “Edit…”
7. Click “New”
8. Enter(with my correct installation)- C:\Temp\qdomyos-zwift\python\x64
- C:\Temp\qdomyos-zwift\python\x64\Scripts\


Again, all of this is for my Win10 device, not the Win11. Please confirm. 




Enter the path to the folder containing the binary you want on your PATH. For example, to add MATLAB R2016b, add:

**Comment by victorypoint** [comment]
Yes, looks good @failedethics. It works on both Win 10 and 11. Note that for step 8, the path is where your QZ is installed.

**Comment by cagnulein** [comment]
@victorypoint the win10 and win11 question was for me because msvc doesn't work on win11 atm.

so it's yes, test it on win10

@victorypoint is it a problem that @failedethics has a 1920 resolution?

**Comment by victorypoint** [comment]
> @victorypoint the win10 and win11 question was for me because msvc doesn't work on win11 atm.
> 
> so it's yes, test it on win10
> 
> @victorypoint is it a problem that @failedethics has a 1920 resolution?

Hmm.. possibly... probably. Have him test the getincline.bat file mentioned in earlier post first to test OCR.

**Comment by failedethics** [comment]
I don't think I'm executing properly, mind giving me the correct text?
![bat](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/cc5c8cd6-fb4f-4e75-b2ee-266a4f1e6a85)

**Comment by victorypoint** [comment]
@failedethics, okay let's double-check a few things. 

1. Where is your QZ installed? If it's installed in say c:\qz for example, then you would add these 2 entries to your system environment PATH variable:
```
- C:\qz\python\x64
- C:\qz\python\x64\Scripts\
```
2. Next, check inside your QZ folder and make sure the python folder is there and it contains am x64 folder which contains all the good python stuff.
3. As stated in the post https://github.com/cagnulein/qdomyos-zwift/issues/1581#issuecomment-1660774335, reboot your PC after adding those items to your system PATH variable.
4. As also stated, run this from the command-line:
```
- python --version
- If installed properly, this will return the python version associated with QZ - Python 3.10.11
```
5. If all good, then run the getincline.bat to test it.

**Comment by failedethics** [comment]
Ok, thanks. 

1. C:\Users\18304\Downloads\windows-msvc2019-binary\release\output
2. Added: 
- C:\Users\18304\Downloads\windows-msvc2019-binary\release\output\qdomyos-zwift\python\x64
- C:\Users\18304\Downloads\windows-msvc2019-binary\release\output\qdomyos-zwift\python\x64\scripts\
3. Rebooted
4. ran: python --version which returned "Python was not found"
5. Attached is python location in qz
![python](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/ac6f425f-fbc1-40e0-ad49-f6e0d821d5e9)

**Comment by cagnulein** [comment]
@failedethics can I suggest to use the same path that @victorypoint described? maybe it's a "" issue on the path, so it's better to use the shorter path as possible

**Comment by failedethics** [comment]
Ok, did that. Nothing. Can I not simply run the python-3.10.11-amd64 executable to install?

**Comment by cagnulein** [comment]
nope because the python version included in qz includes all the dependencies.
we are missing something stupid I guess...could you please post a screenshot of the error just to have all the details?

**Comment by failedethics** [comment]
Ha, ya, well it's probably PEBCAK so I'm just going to include screenshots on everything at this point. 
![pythonerror](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/0e0925a8-603a-4d7b-965e-496d40001842)
![qzloc](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/804cf000-49cf-4038-a4e2-974f17bab0f3)
![path](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/f9c28aac-6478-4c79-a7f1-23d00c275a77)

**Comment by victorypoint** [comment]
Hi @failedethics. Ok good, so your QZ is installed at c:\qz yes? Just confirm that the PATH variables are entered under the system section and not the user section of the environment variables. Reboot. Then also at command-line, type "set" without "s and let us know the output. Then try python --v again to see what happens

**Comment by failedethics** [comment]
@victorypoint oh ya, it was me. Moved to System section and returned "Python 3.10.11" from cmd

**Comment by victorypoint** [comment]
Excellent 👍!! Keep going. Run getincline.bat and hopefully it's good

**Comment by failedethics** [comment]
Ok, can you confirm i'm running the correct command

![qztest](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/1aea8ccb-b86a-4e09-a0c3-67ba01adde78)

**Comment by victorypoint** [comment]
@failedethics, it looks like the spaces in the folder name is causing the problem.

**Comment by failedethics** [comment]
![getincline](https://github.com/cagnulein/qdomyos-zwift/assets/119689504/82258978-2639-4996-bb58-2bc5901fd9f3)

**Comment by victorypoint** [comment]
@failedethics, perfect! You are ready to test auto-incline with QZ and Zwift. Make sure to use Borderless Gaming utility for Zwift. Use Zwift Ultra (1440p) or 4k UHD (2160p) game resolution if you can. Low, Medium, and High won't work well unfortunately.

**Comment by cagnulein** [comment]
@victorypoint maybe the path is something that i can do from qz directly https://stackoverflow.com/questions/17490489/how-can-i-set-path-variable-using-qt ?

**Comment by victorypoint** [comment]
> @victorypoint maybe the path is something that i can do from qz directly https://stackoverflow.com/questions/17490489/how-can-i-set-path-variable-using-qt ?

I was wondering about that to simplify the process. Or I could create a batch file to do it.

**Comment by failedethics** [comment]
[debug-Thu_Sep_28_09_54_10_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12751079/debug-Thu_Sep_28_09_54_10_2023.log)
Zwift set to ultra, windowed. Using borderless gaming.

**Comment by cagnulein** [comment]
always 

python << ERR Traceback (most recent call last):\r\n  File \"C:\\QZ\\zwift-incline.py\", line 92, in <module>\r\n    for word in line:\r\nTypeError: 'NoneType' object is not iterable\r\n"

:( 

i guess we have to debug it dumping a screenshot from the @failedethics setup?

**Comment by victorypoint** [comment]
@cagnulein, let me do a full test with this QZ msvc build and Zwift. I'll try to do that this afternoon.

**Comment by cagnulein** [comment]
thanks Al. unfortunately i don't have a windows pc so i can't help you guys
on this :(

Il giorno gio 28 set 2023 alle 18:44 Al Udell ***@***.***> ha
scritto:

**Comment by victorypoint** [comment]
@cagnulein, no worries. So I compared the msvc build vs mingw build. mingw build is solid and works well as expected. msvc is giving the same python error that @failedethics gets. It seems to be an issue with parsing the None incline value when there is no incline onscreen. When Zwift game shows incline, the value appears to parse correctly.

```
Line   785: Thu Sep 28 19:18:08 2023 1695939488084 Debug: .\windows_zwift_incline_paddleocr_thread.cpp class QString __cdecl windows_zwift_incline_paddleocr_thread::runPython(class QString) run >>  "zwift-incline.py"
	Line   890: Thu Sep 28 19:18:10 2023 1695939490468 Debug: .\bluetooth.cpp void __cdecl bluetooth::debug(const class QString &) "python << OUT "
	Line   891: Thu Sep 28 19:18:10 2023 1695939490468 Debug: .\bluetooth.cpp void __cdecl bluetooth::debug(const class QString &) "python << ERR Traceback (most recent call last):\r\n  File \"c:\\Temp\\qdomyos-zwift\\zwift-incline.py\", line 92, in <module>\r\n    for word in line:\r\nTypeError: 'NoneType' object is not iterable\r\n"
```
	
```
Line  3105: Thu Sep 28 19:18:59 2023 1695939539438 Debug: .\windows_zwift_incline_paddleocr_thread.cpp class QString __cdecl windows_zwift_incline_paddleocr_thread::runPython(class QString) run >>  "zwift-incline.py"
	Line  3318: Thu Sep 28 19:19:04 2023 1695939544039 Debug: .\bluetooth.cpp void __cdecl bluetooth::debug(const class QString &) "python << OUT -1\r\n"
	Line  3319: Thu Sep 28 19:19:04 2023 1695939544039 Debug: .\bluetooth.cpp void __cdecl bluetooth::debug(const class QString &) "python << ERR "
	Line  3329: Thu Sep 28 19:19:04 2023 1695939544140 Debug: .\windows_zwift_incline_paddleocr_thread.cpp class QString __cdecl windows_zwift_incline_paddleocr_thread::runPython(class QString) run >>  "zwift-incline.py"
	Line  3560: Thu Sep 28 19:19:08 2023 1695939548836 Debug: .\bluetooth.cpp void __cdecl bluetooth::debug(const class QString &) "python << OUT -2\r\n"
```

[debug-Thu_Sep_28_19_18_06_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12754403/debug-Thu_Sep_28_19_18_06_2023.log)

**Comment by cagnulein** [comment]
thanks @victorypoint so this means that
1) the msvc is working
2) the @failedethics setup has still something wrong since it seems that paddle can't recognize any pattern on his screen.

How can we debug it?

**Comment by victorypoint** [comment]
> thanks @victorypoint so this means that
> 
> 1. the msvc is working
> 2. the @failedethics setup has still something wrong since it seems that paddle can't recognize any pattern on his screen.
> 
> How can we debug it?

Let me do some more testing and come up with a strategy.

**Comment by failedethics** [comment]
@victorypoint This is probably a silly question but, was there anything with that file that could impact QZs ability to communicate with the treadmill? Because now it isn't picking up the speed as it was previously.

**Comment by failedethics** [comment]
Ok. going through this, the first thing that happened is as soon as QZ connected to the Horizon it changed the incline to 1%(prior to going into Zwift), but no other changes. Other than that, I think I did this correctly. However, Git says the zip is too large to upload here.

**Comment by cagnulein** [comment]
please upload it on google drive or another service
thanks

Il giorno ven 29 set 2023 alle 18:09 failedethics ***@***.***>
ha scritto:

**Comment by failedethics** [comment]
https://drive.google.com/file/d/1drZtgelZX2XorlZhPVUCIqJEFpf6Ggbc/view?usp=sharing

**Comment by victorypoint** [comment]
@failedethics, thank-you for doing this. Looking at the PNG images, I see that QZ remains the foreground window throughout the entire test. Are you using a dual-monitor setup? Zwift needs to be the foreground window for this to work.

**Comment by failedethics** [comment]
Yes, as mentioned [in this comment](https://github.com/cagnulein/qdomyos-zwift/issues/1670#issuecomment-1735937066)

Zwift is running on 1, QZ on 2. Can you ELI5 as to how I ensure it's in the appropriate window, ty.

**Comment by failedethics** [comment]
Nevermind, I switched the windows and incline worked

**Comment by victorypoint** [comment]
> Nevermind, I switched the windows and incline worked

Whoop whoop! Great to hear! Feel free to switch back to the previous zwift-incline.py so you don't fill your hard drive with PNG files.

**Comment by victorypoint** [comment]
@cagnulein, so the only difference at this point between mingw and msvc builds is the python error in msvc when no incline value is onscreen.

`python << ERR Traceback (most recent call last):\r\n File "C:\QZ\zwift-incline.py", line 92, in \r\n for word in line:\r\nTypeError: 'NoneType' object is not iterable\r\n"`

Whereas mingw python shows this:

`Thu Sep 28 19:22:49 2023 1695939769547 Debug: bluetooth.cpp void bluetooth::debug(const QString&) "python << OUT None\r\n"`

**Comment by cagnulein** [comment]
wow great news! i guess that error could be  discarded.

amazing support @victorypoint ! thanks!

Il giorno ven 29 set 2023 alle 19:15 Al Udell ***@***.***> ha
scritto:

**Comment by failedethics** [comment]
Thank you so much for all the help!!

A few follow-ups:
Is the incline value supposed to be 1:1? Ex: Incline in Zwift is 6%, should the value on the horizon be 6%?
How long should it take to register the incline value? There were times of up to a 10 second delay for QZ to recognize the new incline value in Zwift. 
During the workout, Zwift calls out speed settings, can QZ auto-adjust the speeds as well? 
Latest test attached.

https://drive.google.com/file/d/1Kx0KGMfFAPdQ-XWbLC6LMAW3-pgDaZ9c/view?usp=sharing

**Comment by victorypoint** [comment]
@failedethics, in a perfect world OCR incline detection would be 1:1, unfortunately we're working hard to speed up the incline detection. Your test post shows that your PC took 3.86 seconds to detect an incline. On my PC with a newer Intel CPU the detection is much quicker. The quicker the detection obviously the quicker your treadmill responds.

Regarding your workout question, yes this works in QZ, you just need to toggle on the Zwift workout option, not the Zwift incline option. Your speed and incline will be obtained from the onscreen workout instruction.

**Comment by failedethics** [comment]
Awesome. Building a new Zwift pc soon based on ZPCMR recommendations. Any
additional specs you recommend to get the best from QZ?

**Comment by victorypoint** [comment]
> Awesome. Building a new Zwift pc soon based on ZPCMR recommendations. Any additional specs you recommend to get the best from QZ?

Excellent! Just follow Dave Higgins recommendations in the files section. I did a budget build and it's working great! For CPU I chose the Core i3-12100F. The 12100F features 4 cores and 8 threads with great single threaded performance and includes a cooler in the box. I went with 16GB of 3200MHz DDR4 RAM and an ASRock H610M-ITX/ac motherboard. This was a few years ago so the recommended specs have probably changed a bit.

**Comment by failedethics** [comment]
Awesome, appreciate the feedback! 

One last thing, I replaced the zwift-incline.py with the original, tested the "Zwift Treadmill Auto Workout" and got no speed AND no incline. Are there any other settings that need to be enabled? 
[debug-Fri_Sep_29_12_51_16_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/12769890/debug-Fri_Sep_29_12_51_16_2023.log)

**Comment by victorypoint** [comment]
> One last thing, I replaced the zwift-incline.py with the original, tested the "Zwift Treadmill Auto Workout" and got no speed AND no incline. Are there any other settings that need to be enabled? [debug-Fri_Sep_29_12_51_16_2023.log]

Very strange. I don't see any inclines received in your QZ log. There's no difference between the 2 files other than exporting PNG files for debugging. Reboot treadmill? Sounds like the QZ/treadmill bluetooth link didn't happen.
